lib/datasource/go/goproxy.ts

```diff
@@ -0,0 +1,189 @@
+    .split(/([^,|]*(?:,|\|))/)
+    .filter(Boolean)
+    .map((s) => s.split(/(?=,|\|)/))
+    .map(([url, separator]) => ({
+      url,
+      fallback:
+        separator === ','
+          ? GoproxyFallback.WhenNotFoundOrGone
+          : GoproxyFallback.Always,
+    }));
+
+  // Ignore hosts after any keyword
+  for (let idx = 0; idx < result.length; idx += 1) {
+    const { url } = result[idx];
+    if (['off', 'direct'].includes(url)) {
+      result = result.slice(0, idx);
+      break;
+    }
+  }
+
+  return result;
+}
+
+// https://golang.org/pkg/path/#Match
+const lexer = moo.states({
+  main: {
+    separator: {
+      match: /\s*?,\s*?/,
+      value: (_: string) => '|',
+    },
+    asterisk: {
+      match: '*',
+      value: (_: string) => '[^\\/]*',
+    },
+    qmark: {
+      match: '?',
+      value: (_: string) => '[^\\/]',
+    },
+    characterRangeOpen: {
+      match: '[',
+      push: 'characterRange',
+      value: (_: string) => '[',
+    },
+    char: /[^*?\\[\n]/,
+    escapedChar: {
+      match: /\\./,
+      value: (s: string) => s.slice(1),
+    },
+  },
+  characterRange: {
+    char: /[^\\\]\n]/,
+    escapedChar: {
+      match: /\\./,
+      value: (s: string) => s.slice(1),
+    },
+    characterRangeEnd: {
+      match: ']',
+      pop: 1,
+    },
+  },
+});
+
+export function parseNoproxy(
+  input: unknown = process.env.GONOPROXY || process.env.GOPRIVATE
+): RegExp | null {
+  if (!is.string(input)) {
+    return null;
+  }
+  lexer.reset(input);
+  const noproxyPattern = [...lexer].map(({ value }) => value).join('');
+  return noproxyPattern ? regEx(`^(?:${noproxyPattern})$`) : null;
+}
+
+/**
+ * Avoid ambiguity when serving from case-insensitive file systems.
+ *
+ * @see https://golang.org/ref/mod#goproxy-protocol
+ */
+export function encodeCase(input: string): string {
+  return input.replace(/([A-Z])/g, (x) => `!${x.toLowerCase()}`);
+}
+
+export async function listVersions(
+  baseUrl: string,
+  lookupName: string
+): Promise<string[]> {
+  const url = `${baseUrl}/${encodeCase(lookupName)}/@v/list`;
+  const { body } = await http.get(url);
+  return body
+    .split(/\s+/)
+    .filter(Boolean)
+    .filter((x) => x.indexOf('+') === -1);
+}
+
+export async function versionInfo(
+  baseUrl: string,
+  lookupName: string,
+  version: string
+): Promise<Release> {
+  const url = `${baseUrl}/${encodeCase(lookupName)}/@v/${version}.info`;
+  const res = await http.getJson<VersionInfo>(url);
+
+  const result: Release = {
+    version: res.body.Version,
+  };
+
+  if (res.body.Time) {
+    result.releaseTimestamp = res.body.Time;
+  }
+
+  return result;
+}
+
+export async function getReleases(
+  config: GetReleasesConfig
+): Promise<ReleaseResult | null> {
+  const { lookupName } = config;
+
+  const noproxy = parseNoproxy();
+  if (noproxy?.test(lookupName)) {
+    logger.debug(`Skipping ${lookupName} via GONOPROXY match`);
+    return null;
+  }
+
+  const proxyList = parseGoproxy();
```

Comment:
Shouldn't we cache this? i don't think the env will change at runtime. 🤔

Reply:
We should one day allow it to be user configurable but for now it's static